tests/test_TFNetworkLayer.py

```diff
@@ -331,10 +362,440 @@ def test_batch_norm_unequal_seq_len():
+          output=BatchNormLayer.get_out_data_from_opts(
+            name="batch_norm",
+            sources=[src_nhwc],
+            network=net))
+        batch_norm.post_init(layer_desc=None)
+      tf.global_variables_initializer().run()
+      # Save all variables before changing them:
+      with tf.name_scope("saver"):
+        saver = tf.train.Saver(var_list=batch_norm.params, max_to_keep=2 ** 31 - 1)
+        saver.save(session, "/tmp/model.ckpt")
+      # Call will modify the variables, but this won't be saved
+      out = session.run(
+        batch_norm.output.placeholder,
+        feed_dict={src_nhwc.output.placeholder: data, src_nhwc.output.size_placeholder[0]: seq_lens})
+      bn_vars = {}
+      for k, v in batch_norm.params.items():
+        # Evaluate vars after update
+        bn_vars[k] = v.eval()
+      return out, bn_vars
+
+  def load_nhwc_vars_to_nchw_bn(data, seq_lens):
+    # Create nchw batch_norm layer and retore variables from the checkpoint, which was created
+    # with a nhwc batch_norm layer
+    with make_scope() as session:
+      net = TFNetwork(extern_data=ExternData())
+      net.train_flag = True
+      with tf.variable_scope("src_nchw"):
+        src_nchw = InternalLayer(name="src_nchw", network=net, out_type={
+            "dim": 16,
+          "shape": (16, None, 16),
+          "batch_dim_axis": 0,
+          "time_dim_axis": 2,
+          "feature_dim_axis": 1,
+          "sparse": False})
+        src_nchw.output.placeholder = tf.placeholder(shape=(None, 16, None, 16), dtype=tf.float32)
+        src_nchw.output.size_placeholder = {1: tf.placeholder(shape=(None,), dtype=tf.int32)}
+
+      with tf.variable_scope("batch_norm"):
+        batch_norm = BatchNormLayer(
+          name="batch_norm", network=net, masked_time=False, sources=[src_nchw],
+          output=BatchNormLayer.get_out_data_from_opts(
+            name="batch_norm",
+            sources=[src_nchw],
+            network=net))
+        batch_norm.post_init(layer_desc=None)
+      # Restore the variables
+      with tf.name_scope("saver"):
+        saver = tf.train.Saver(var_list=batch_norm.params, max_to_keep=2 ** 31 - 1)
+        saver.restore(session, "/tmp/model.ckpt")
+      # Call will modify the variables.
+      # The output and the variables have to be the same as in the nhwc case
+      out = session.run(
+        batch_norm.output.placeholder,
+        feed_dict={src_nchw.output.placeholder: data, src_nchw.output.size_placeholder[1]: seq_lens})
+      bn_vars = {}
+      for k, v in batch_norm.params.items():
+        # Evaluate vars after update
+        bn_vars[k] = v.eval()
+      return out, bn_vars
+
+  rnd = np.random.RandomState(42)
+  input_data = rnd.rand(10, 11, 16, 16)
+  input_data_nchw = np.transpose(input_data, axes=(0, 3, 1, 2))
+  # Make sure that the data is equal
+  assert np.array_equal(input_data[:, :, :, 2], input_data_nchw[:, 2, :, :])
+  seq_lens = np.array([11, 11, 11, 11, 11, 11, 11, 11, 11, 11])
```

Comment:
As I wrote before, you should not use this case where all seq lens are the same. You are hiding many bugs by that.